### ISSUE TITLE
Missing Authentication header when using with OpenRouter.ai

### ISSUE THREAD

**Issue by kirilligum** [paul-gauthier/aider#297]
when i put open router arguments just like in [the faq](https://aider.chat/docs/faq.html#accessing-other-llms-with-openrouter) in command line, i get the following error:

```
(main) [1]> aider –-openai-api-base https://openrouter.ai/api/v1 -–openai-api-key sk-or-v1-xxxxx -–model anthropic/claude-2
usage: aider [-h] [--openai-api-key OPENAI_API_KEY] [--model MODEL] [--skip-model-availability-check SKIP_MODEL_AVAILABILITY_CHECK] [-3] [--voice-language VOICE_LANGUAGE] [--openai-api-base OPENAI_API_BASE] [--openai-api-type OPENAI_API_TYPE]
             [--openai-api-version OPENAI_API_VERSION] [--openai-api-deployment-id OPENAI_API_DEPLOYMENT_ID] [--openai-api-engine OPENAI_API_ENGINE] [--edit-format EDIT_FORMAT] [--map-tokens MAP_TOKENS] [--input-history-file INPUT_HISTORY_FILE]
             [--chat-history-file CHAT_HISTORY_FILE] [--dark-mode] [--light-mode] [--pretty | --no-pretty] [--stream | --no-stream] [--user-input-color USER_INPUT_COLOR] [--tool-output-color TOOL_OUTPUT_COLOR] [--tool-error-color TOOL_ERROR_COLOR]
             [--assistant-output-color ASSISTANT_OUTPUT_COLOR] [--code-theme CODE_THEME] [--show-diffs] [--git | --no-git] [--gitignore | --no-gitignore] [--aiderignore AIDERIGNORE] [--auto-commits | --no-auto-commits]
             [--dirty-commits | --no-dirty-commits] [--dry-run | --no-dry-run] [--commit] [--version] [--apply FILE] [--yes] [-v] [--show-repo-map] [--message COMMAND] [--encoding ENCODING] [-c CONFIG_FILE]
             [FILE ...]
aider: error: unrecognized arguments: -–openai-api-key sk-or-v1-xxxxx -–model anthropic/claude-2
```

so i tried to put the variables into the .aider.conf.yml file

```
(main)> cat .aider.conf.yml
openai-api-key: sk-or-v1-xxxxx
openai-api-base: https://openrouter.ai/api/v1
model: gpt-4
```

then i run aider and get an error

```
(main)> aider
Aider v0.15.0
Setting openai.api_base=https://openrouter.ai/api/v1
Model: openai/gpt-4
Git repo: .git
Repo-map: basic using 1024 tokens (ctags --version does not claim to be universal ctags)
Use /help to see in-chat commands, run with --help to see cmd line args
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
> create nextjs example

Traceback (most recent call last):
  File "/home/kirill/.local/bin/aider", line 8, in <module>
    sys.exit(main())
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/main.py", line 564, in main
    coder.run()
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 326, in run
    new_user_message = self.run_loop()
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 409, in run_loop
    return self.send_new_user_message(inp)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 442, in send_new_user_message
    interrupted = self.send(messages, functions=self.functions)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 562, in send
    hash_object, completion = send_with_retries(model, messages, functions, self.stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/sendchat.py", line 64, in send_with_retries
    res = openai.ChatCompletion.create(**kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create
    response, _, api_key = requestor.request(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 298, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 700, in _interpret_response
    self._interpret_response_line(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 763, in _interpret_response_line
    raise self.handle_error_response(
openai.error.AuthenticationError: Missing Authentication header
```

**any ideas on how to resolve this? is this openrouter problem by any chance?**

open router runs fine with cli:

```
(main) [1]> curl https://openrouter.ai/api/v1/chat/completions \
                                                                                  -H "Content-Type: application/json" \
                                                                                  -H "Authorization: Bearer sk-or-v1-xxxxx" \
                                                                                  -H "HTTP-Referer:  http://localhost:3000/" \
                                                                                  -H "X-Title: clitest" \
                                                                                  -d '{
                                                                              "model": "openai/gpt-4-32k",
                                                                              "messages": [
                                                                                {"role": "user", "content": "What is the meaning of life?"}
                                                                              ]
                                                                            }'

{"choices":[{"message":{"role":"assistant","content":"As an artificial intelligence, I don't have personal experiences or beliefs. However, I can share that in philosophy, religion, and sociology, the meaning of life is considered to be personal and subjective, and often involves concepts of happiness, fulfillment, love, and the pursuit of knowledge. Different individuals and societies may have different beliefs about the purpose and meaning of life."},"finish_reason":"stop"}],"model":"gpt-4-32k-0613","usage":{"prompt_tokens":14,"completion_tokens":73,"total_tokens":87},"id":"gen-ypGUreGrNP4ugxMp4qqWGVkgwQwE","object":"chat.completion","created":1698386676}⏎
```

**Comment by markenzie** [comment]
I'm not sure what's going on here but I can reproduce this.

You can solve it by just passing the --openai-api-base and --openai-api-key arguments manually or set those as environment vars (in ENV_VAR format) and it will work.

**Comment by kirilligum** [comment]
thank you. i set AIDER_MODEL, AIDER_OPENAI_API_KEY, and AIDER_OPENAI_API_BASE env vars and it worked

**Comment by Korolen** [comment]
I'm still getting this error no matter how I pass in the https://openrouter.ai/api/v1 base and openrouter.ai api key.

**Comment by kirilligum** [comment]
Yeah, I still get an error:

```
kirill@zagadka ~/p/prompt_eng_market (nextjs)> aider --version
aider 0.16.1-dev
kirill@zagadka ~/p/prompt_eng_market (nextjs)> set -x AIDER_MODEL anthropic/claude-2
kirill@zagadka ~/p/prompt_eng_market (nextjs)> set -x AIDER_OPENAI_API_KEY sk-or-v1-xxxxx
6a9a064e46afb1e9563427e
kirill@zagadka ~/p/prompt_eng_market (nextjs)> set -x AIDER_OPENAI_API_BASE https://openrouter.ai/api/v1
kirill@zagadka ~/p/prompt_eng_market (nextjs)> aider -m hi
Aider v0.16.1-dev
Setting openai.api_base=https://openrouter.ai/api/v1
Model: anthropic/claude-2
Git repo: .git
Repo-map: using 1024 tokens
Use /help to see in-chat commands, run with --help to see cmd line args

Traceback (most recent call last):
  File "/home/kirill/.local/bin/aider", line 8, in <module>
    sys.exit(main())
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/main.py", line 562, in main
    coder.run(with_message=args.message)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 322, in run
    new_user_message = self.send_new_user_message(new_user_message)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 435, in send_new_user_message
    interrupted = self.send(messages, functions=self.functions)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 555, in send
    hash_object, completion = send_with_retries(model, messages, functions, self.stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/sendchat.py", line 64, in send_with_retries
    res = openai.ChatCompletion.create(**kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in
create
    response, _, api_key = requestor.request(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 230, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 624, in _interpret_response
    self._interpret_response_line(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 687, in _interpret_response_line
    raise self.handle_error_response(
openai.error.AuthenticationError: Missing Authentication header
```

the same with Aider v0.16.0

**Comment by kirilligum** [comment]
@markenzie  passing the variables didn't work. i tried it 3 different ways. you can see above

**Comment by joshuavial** [comment]
I think the error is with the model - try `openai/gpt-4`

**Comment by joshuavial** [comment]
@kirilligum I'm not sure why this is happening. I ran the following against 0.16.1 and I can't see what is different between this and what you are doing. It looks like your base and model are set correctly.

your api key is correct or it would show an invalid credentials error. This error would show up if line 54 of `aider/sendchat.py` wasn't being hit

```
   if "openrouter.ai" in openai.api_base:
        kwargs["headers"] = {"HTTP-Referer": "http://aider.chat", "X-Title": "Aider"}

```

I see you're running off the dev branch, can you put in a `print(openai.api_base` in on line 52 of that file?

This is what I see when testing locally.
```
 aider  --openai-api-base=https://openrouter.ai/api/v1 --openai-api-key='sk-or-v1-.....'  --model=anthropic/claude-2
Aider v0.16.1
Add .aider* to .gitignore (recommended)? y
Added .aider* to .gitignore
Setting openai.api_base=https://openrouter.ai/api/v1
Model: anthropic/claude-2
Git repo: .git
Repo-map: using 1024 tokens
Use /help to see in-chat commands, run with --help to see cmd line args
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────> create hello world inpython

To create a simple "Hello World" program in Python, I would suggest the following:


 new_file.py
 <<<<<<< SEARCH

 =======

 print("Hello World!")

 >>>>>>> REPLACE


This uses a SEARCH/REPLACE block to add a new file called new_file.py with a print statement that outputs "Hello World!". The SEARCH section is left
empty since this is a new file.

Let me know if you would like me to explain or modify the changes in any way. I'm happy to provide more details on how this Python program works.

Allow creation of new file new_file.py? y
Applied edit to new_file.py
Commit 6e0efe8 aider: Added new_file.py with a print statement that outputs "Hello World!".
```

**Comment by kirilligum** [comment]
thank you @joshuavial , single quotes in the command line worked.

i opened a pr to add it to the faq https://github.com/paul-gauthier/aider/pull/311

however setting the env (fish) variable with or without quotes didn't work. 

```
kirill@zagadka ~/p/prompt_eng_market (nextjs) [2]> set -e AIDER_OPENAI_API_KEY
kirill@zagadka ~/p/prompt_eng_market (nextjs) [4]> set -x AIDER_OPENAI_API_KEY 'sk-or-v1-xxxxx'
kirill@zagadka ~/p/prompt_eng_market (nextjs) [4]> aider -m hey
Aider v0.16.2
Setting openai.api_base=https://openrouter.ai/api/v1
Model: anthropic/claude-2
Git repo: .git
Repo-map: using 1024 tokens
Use /help to see in-chat commands, run with --help to see cmd line args

Traceback (most recent call last):
  File "/home/kirill/.local/bin/aider", line 8, in <module>
    sys.exit(main())
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/main.py", line 562, in main
    coder.run(with_message=args.message)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 322, in run
    new_user_message = self.send_new_user_message(new_user_message)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 435, in send_new_user_message
    interrupted = self.send(messages, functions=self.functions)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/coders/base_coder.py", line 555, in send
    hash_object, completion = send_with_retries(model, messages, functions, self.stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/aider/sendchat.py", line 64, in send_with_retries
    res = openai.ChatCompletion.create(**kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in
create
    response, _, api_key = requestor.request(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 230, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 624, in _interpret_response
    self._interpret_response_line(
  File "/home/kirill/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 687, in _interpret_response_line
    raise self.handle_error_response(
openai.error.AuthenticationError: Missing Authentication header
kirill@zagadka ~/p/prompt_eng_market (next) [1]>
```

modifying .aider.conf.yml also didn't work with single or double quotes:
```
kirill@zagadka ~/p/prompt_eng_market (nextjs) [1]> cat .aider.conf.yml
openai-api-key: "sk-or-v1-xxxxx"
openai-api-base: https://openrouter.ai/api/v1
model: mistralai/mistral-7b-instruct
```

**Comment by kirilligum** [comment]
in bash, worked without single quotes

**Comment by Korolen** [comment]
I fixed my problem .. in my ~/.secrets file my openrouter key somehow lost its "-or-v1" snippet. Was probably my own fault somewhere along the line. _So, it appears this may be interpretable as a generic "key incorrect" error?_

It's possible upgrading to aider 0.16.3 was an additional help in the background as well while I diagnosed the key issue.

Anyway, I am now able to successfully launch aider w/ openrouter again using both environment variables (AIDER_MODEL, AIDER_OPENAI_API_BASE, OPENAI_API_KEY is the trio), or command-line args (--model, --openai-api-base, --openai-api-key), with or without equals and with or without single or double quotes. In other words, differentiating between any of these methods of passing in the information didn't change my successful result once I fixed my key.

**Comment by joshuavial** [comment]
Interesting, the first thing I tested was throwing a 'sk-or-v1-bananas' api key at openrouter to see if that was the issue and it gave me a completely different error (along the lines of incorrect key). Perhaps they return a different error if the key is completely malformed vs just off a little.  Either way it sounds like the code could do with a little hardening and better error message.

One issue I've seen in the past is that if OPENAI_API_KEY is set in the environment it will override the contents of an .aider.conf file, but not the command line args. So if I'm using a .conf file I had to call `unset OPENAI_API_KEY` before it would be happy. I've switched to using command line args and aliases instead of the .conf file approach personally.
